lib/internal/event_target.js

```diff
@@ -0,0 +1,392 @@
+
+  const callback =
+    typeof listener === 'function' ?
+      listener.bind(this) :
+      listener.handleEvent.bind(listener);
+
+  const handler = {
+    callback,
+    once,
+    remove: removeEventListener.bind(this, type, listener)
+  };
+
+  handlers.set(listener, handler);
+}
+
+function removeEventListener(type, listener) {
+  const state = targets.get(this);
+  if (state === undefined)
+    throw new ERR_INVALID_THIS('EventTarget');
+
+  validateType(type);
+  validateListener(listener);
+
+  const handlers = state.events.get(type);
+  if (handlers !== undefined) {
+    handlers.delete(listener);
+    if (handlers.size === 0)
+      state.events.delete(type);
+  }
+}
+
+function addCatch(that, promise, event) {
+  const then = promise.then;
+  if (typeof then === 'function') {
+    then.call(promise, undefined, function(err) {
+      // The callback is called with nextTick to avoid a follow-up
+      // rejection from this promise.
+      process.nextTick(emitUnhandledRejectionOrErr, that, err, event);
+    });
+  }
+}
+
+function emitUnhandledRejectionOrErr(that, err, event) {
+  if (typeof that[kRejection] === 'function') {
+    that[kRejection](err, event);
+  } else {
+    process.emit('uncaughtException', err, event);
+  }
+}
+
+function dispatchEvent(event) {
+  const state = targets.get(this);
+  if (state === undefined)
+    throw new ERR_INVALID_THIS('EventTarget');
+
+  if (event == null ||
+      typeof event !== 'object' ||
+      typeof event.type !== 'string') {
+    throw new ERR_INVALID_ARG_TYPE('event', 'Event', event);
+  }
+
+  if (state.emitting.has(event.type)) {
+    throw new ERR_EVENT_RECURSION(event.type);
+  }
+
+  const handlers = state.events.get(event.type);
+  if (handlers !== undefined) {
+    state.emitting.add(event.type);
+    try {
+      const copy = ArrayFrom(handlers.values());
+      for (let n = 0; n < copy.length; n++) {
+        if (copy[n].once)
+          copy[n].remove();
+        try {
+          const result = copy[n].callback(event);
+          if (result !== undefined && result !== null) {
+            addCatch(this, result, event);
+          }
+        } catch (err) {
+          emitUnhandledRejectionOrErr(this, err, event);
+        }
+      }
+    } finally {
+      state.emitting.delete(event.type);
+    }
+  }
+}
+
+// EventEmitter-ish API:
+
+function eventNames() {
+  const state = targets.get(this);
+  if (state === undefined)
+    throw new ERR_INVALID_THIS('EventTarget');
+  return ArrayFrom(state.events.keys());
+}
+
+function listenerCount(type) {
+  const state = targets.get(this);
+  if (state === undefined)
+    throw new ERR_INVALID_THIS('EventTarget');
+  validateType(type);
+  const handlers = state.events.get(type);
+  return handlers !== undefined ? handlers.size : 0;
+}
+
+function off(type, listener) {
+  this.removeEventListener(type, listener);
+  return this;
+}
+
+function on(type, listener) {
+  this.addEventListener(type, listener);
+  return this;
+}
```

Comment:
Posing this as an open question: `on`/`addListener` on `EventEmitter` behaves different to `addEventListener` with regard to the same listener being added twice. How should this be handled?

- Differ in behavior from `EventEmitter` (current state of the PR)?
- Differ in behavior depending on which method is used?

Reply:
My preference would be to differ in behavior from `EventEmitter`. Consumers of this API are already going to have to treat it differently from `EventEmitter` due to the fact that handler functions receive only a single `event` argument.

Reply:
Another option is to behave like EventTarget but emit a warning if the same listener is added more than once.

I suspect that it's quite surprising/confusing behavior and it's rarely intentional (assuming someone using the node name here)

Reply:
For this, I'd suspect that anyone using the `EventTarget` interface is likely going to be most familiar with the register-only-once behavior currently implemented by browsers

Reply:
@jasnell I am _pretty_ confident the vast majority of users don't knows of the `target.addEventListener(fn); target.addEventListener(fn);` behavior (and certainly don't rely on it). 

While in _theory_ this can cause some terrible bugs in practice I suspect that it really won't. So I don't feel strongly enough to argue that point though so I am fine with either of the three options (behave like EventTarget, behave like EventEmitter, behave like either and warn).

Reply:
The behave like `EventTarget` implementation is *far* simpler so unless we get someone specifically requesting to be able to add a single handler multiple times (and has a good reason for doing so), let's keep it the way it is.